narwhal/primary/src/primary.rs

```diff
@@ -547,6 +542,223 @@ impl PrimaryReceiverHandler {
+                    ensure!(
+                        narwhal_round <= header.round,
+                        DagError::TooOld(header.digest().into(), header.round, narwhal_round)
+                    )
+                },
+            }
+        }
+
+        // Ensure we have the parents. If any are missing, the requester should provide them on retry.
+        let (parents, missing) = self.synchronizer.get_parents(header)?;
+        if !missing.is_empty() {
+            return Ok(RequestVoteResponse {
+                vote: None,
+                missing,
+            });
+        }
+
+        // Now that we've got all the required certificates, ensure we're voting on a
+        // current Header.
+        narwhal_round = *rx_narwhal_round_updates.borrow();
+        ensure!(
+            narwhal_round <= header.round,
+            DagError::TooOld(header.digest().into(), header.round, narwhal_round)
+        );
+
+        // Check the parent certificates. Ensure the parents:
+        // - form a quorum
+        // - are all from the previous round
+        // - are from unique authorities
+        let mut parent_authorities = BTreeSet::new();
+        let mut stake = 0;
+        for parent in parents.iter() {
+            ensure!(
+                parent.round() + 1 == header.round,
+                DagError::MalformedHeader(header.digest())
+            );
+            ensure!(
+                parent_authorities.insert(&parent.header.author),
+                DagError::MalformedHeader(header.digest())
+            );
+            stake += committee.stake(&parent.origin());
+        }
+        ensure!(
+            stake >= committee.quorum_threshold(),
+            DagError::HeaderRequiresQuorum(header.digest())
+        );
+
+        // Synchronize all batches referenced in the header.
+        self.synchronizer
+            .sync_batches(header, network, /* max_age */ 0)
+            .await?;
+
+        // Store the header.
+        self.header_store
+            .async_write(header.digest(), header.clone())
+            .await;
+
+        // Check if we can vote for this header.
+        // Send the vote when:
+        // 1. when there is no existing vote for this publicKey & epoch/round
+        // 2. when there is a vote for this publicKey & epoch/round, and the vote is the same
+        // Taking the inverse of these two, the only time we don't want to vote is when:
+        // there is a digest for the publicKey & epoch/round, and it does not match the digest
+        // of the vote we create for this header.
+        // Also when the header is older than one we've already voted for, it is useless to vote,
+        // so we don't.
+        let result = self
+            .vote_digest_store
+            .read(header.author.clone())
+            .await
+            .map_err(DagError::StoreError)?;
+
+        if let Some(vote_info) = result {
+            if header.epoch < vote_info.epoch
```

Comment:
Would this also want to be `header.epoch != vote_info.epoch`?

Reply:
I think we'll have to re-evaluate epoch handling in narwhal generally since a lot of it probably doesn't work correctly. but my thinking here was that there's no reason to withhold the vote if all checks pass just b/c some local epoch state maybe hasn't advanced yet.

Reply:
(In this case if we made it `!=`, we'd actually refuse to ever vote on an epoch newer than the one we started with, so I think that wouldn't be right either.)

Reply:
Yeah i do think that the whole handling of epoch changes needs to be done holistically, because without a formalized plan, i do not know what the correct think to do here is.

Reply:
I look forward to just deleting most/all epoch related code from NW :)